src/atoms/pushNotification.ts

```diff
@@ -0,0 +1,200 @@
+import messaging, {FirebaseMessagingTypes} from '@react-native-firebase/messaging';
+import {atom, RecoilState, useRecoilState, useRecoilValue, useSetRecoilState} from 'recoil';
+import {persistAtom} from '@atoms/persist';
+import {useCallback, useEffect, useMemo, useState} from 'react';
+import {useAppState} from 'src/hooks/useAppState';
+
+type AuthorizationStatus = FirebaseMessagingTypes.AuthorizationStatus;
+
+enum SUBSCRIPTION_ID {
+  NEW_TREASURY_PROPOSAL = 'treasury.Proposed',
+  TIP_SUGGESTION = 'tips.NewTip',
+  NEW_REFERENDUM = 'democracy.Started',
+}
+
+type TogglePushTopicPayload = {
+  id: SUBSCRIPTION_ID;
+  subscribe: boolean;
+};
+
+type Subscription = {
+  [key in SUBSCRIPTION_ID]: {isSubscribed: boolean};
+};
+
+const SUBSCRIPTION_INITIAL_STATE = {
+  [SUBSCRIPTION_ID.NEW_TREASURY_PROPOSAL]: {isSubscribed: false},
+  [SUBSCRIPTION_ID.TIP_SUGGESTION]: {isSubscribed: false},
+  [SUBSCRIPTION_ID.NEW_REFERENDUM]: {isSubscribed: false},
+};
+
+const SUBSCRIPTION_LABEL = {
+  [SUBSCRIPTION_ID.NEW_TREASURY_PROPOSAL]: 'New Treasury Proposal',
+  [SUBSCRIPTION_ID.TIP_SUGGESTION]: 'Tip Suggestion',
+  [SUBSCRIPTION_ID.NEW_REFERENDUM]: 'New Referendum',
+};
+
+const subscriptionState: RecoilState<Subscription> = atom({
+  key: 'pushNotificationSubscription',
+  default: SUBSCRIPTION_INITIAL_STATE,
+  effects: [persistAtom],
+});
+
+const isPermissionGranted = (status: AuthorizationStatus) =>
+  status === messaging.AuthorizationStatus.AUTHORIZED || status === messaging.AuthorizationStatus.PROVISIONAL;
+
+const isPermissionDenied = (status: AuthorizationStatus) => status === messaging.AuthorizationStatus.DENIED;
+
+const isPermissionNotDetermined = (status: AuthorizationStatus) =>
+  status === messaging.AuthorizationStatus.NOT_DETERMINED;
+
+export function usePushTopics() {
+  const [subscriptionTopics, setSubscriptionTopics] = useRecoilState(subscriptionState);
+
+  const topics = useMemo(() => {
+    const topicIds = Object.keys(subscriptionTopics) as SUBSCRIPTION_ID[];
+    return topicIds.map((topicId) => ({
+      id: topicId,
+      isSubscribed: subscriptionTopics[topicId].isSubscribed,
+      label: SUBSCRIPTION_LABEL[topicId],
+    }));
+  }, [subscriptionTopics]);
+
+  const toggleTopic = useCallback(
+    ({id, subscribe}: TogglePushTopicPayload) => {
+      subscribe ? messaging().subscribeToTopic(id) : messaging().unsubscribeFromTopic(id);
+      setSubscriptionTopics({
+        ...subscriptionTopics,
+        [id]: {isSubscribed: subscribe},
+      });
+    },
+    [subscriptionTopics, setSubscriptionTopics],
+  );
+
+  const subscribeToAllTopics = useCallback(() => {
+    const ids = Object.keys(subscriptionTopics) as SUBSCRIPTION_ID[];
+    const subscriptionPromises = ids.map((topicId) => messaging().subscribeToTopic(topicId));
+
+    Promise.all(subscriptionPromises);
+
+    const subscribedTopics = ids.reduce((subscription, topicId) => {
```

Comment:
We probably should not assume what all requests to subscribeToTopic will succeed, as they can fail for multiple reasons (e.g: token has expired or network issue). It will be safer to await for `Promise.allSettled` and mark them as true for each fulfilled promise.

Consider the same for single topic subscriptions